src/App/src/Middleware/DeprecationMiddleware.php

```diff
@@ -0,0 +1,116 @@
+        if (! $routeResult instanceof RouteResult || $routeResult->isFailure()) {
+            return $response;
+        }
+
+        $reflectionHandler = null;
+        $routeMiddleware   = $routeResult->getMatchedRoute()->getMiddleware();
+        if ($routeMiddleware instanceof LazyLoadingMiddleware) {
+            $reflectionMiddlewareClass = new ReflectionClass($routeMiddleware->middlewareName);
+            if ($reflectionMiddlewareClass->implementsInterface(RequestHandlerInterface::class)) {
+                $reflectionHandler = $reflectionMiddlewareClass;
+            }
+        } else {
+            $reflectionClass    = new ReflectionClass($routeMiddleware);
+            $middlewarePipeline = $reflectionClass->getProperty('pipeline')->getValue($routeMiddleware);
+            for ($middlewarePipeline->rewind(); $middlewarePipeline->valid(); $middlewarePipeline->next()) {
+                $reflectionMiddlewareClass = new ReflectionClass($middlewarePipeline->current()->middlewareName);
+                if ($reflectionMiddlewareClass->implementsInterface(RequestHandlerInterface::class)) {
+                    $reflectionHandler = $reflectionMiddlewareClass;
+                }
+            }
+        }
+
+        if (! $reflectionHandler) {
+            return $response;
+        }
+
+        $attributes = $this->getAttributes($reflectionHandler, self::RESOURCE_DEPRECATION_ATTRIBUTE);
+        foreach ($reflectionHandler->getMethods() as $method) {
+            $methodRef   = new ReflectionMethod($method->class, $method->name);
+            $attributes += $this->getAttributes($methodRef, self::METHOD_DEPRECATION_ATTRIBUTE);
+        }
+
+        if ([self::RESOURCE_DEPRECATION_ATTRIBUTE, self::METHOD_DEPRECATION_ATTRIBUTE] === array_keys($attributes)) {
+            throw new DeprecationConflictException(
+                sprintf(
+                    Message::RESTRICTION_DEPRECATION,
+                    self::RESOURCE_DEPRECATION_ATTRIBUTE,
+                    self::METHOD_DEPRECATION_ATTRIBUTE
+                )
+            );
+        }
+
+        $sunset = '';
+        $link   = '';
+        if ($attributes[self::RESOURCE_DEPRECATION_ATTRIBUTE] ?? '') {
+            $sunset = $attributes[self::RESOURCE_DEPRECATION_ATTRIBUTE]['sunset'];
+            $link   = $attributes[self::RESOURCE_DEPRECATION_ATTRIBUTE]['link'];
+        }
+
+        if ($attributes[self::METHOD_DEPRECATION_ATTRIBUTE] ?? '') {
+            $sunset = $attributes[self::METHOD_DEPRECATION_ATTRIBUTE]['sunset'];
+            $link   = $attributes[self::METHOD_DEPRECATION_ATTRIBUTE]['link'];
+        }
+
+        if ($sunset) {
+            $response = $response->withHeader('sunset', $sunset);
+        }
+
+        if ($link) {
+            $response = $response->withHeader('link', $link);
+        }
+
+        return $response;
+    }
+
+    public function getAttributes(ReflectionClass|ReflectionMethod $reflection, string $type): array
```

Comment:
```suggestion
    /**
     * @param class-string $type
     */
    public function getAttributes(ReflectionClass|ReflectionMethod $reflection, string $type): array
```

---

To fix this Psalm error:

> ArgumentTypeCoercion: Argument 1 of ReflectionClass::getAttributes expects class-string|null, but parent type string provided